R/bambu-processReads_scoreReadClasses.R

```diff
@@ -151,63 +157,73 @@ prepareGeneModelFeatures = function(rowData){
         # or compat with only 1 but are not equal
         numNonSubsetRCs = numRCs-sum(compatible>=2 | (compatible==1 & !equal)),
         highConfidence=any(confidenceType=='highConfidenceJunctionReads')) %>%
-    mutate(numReads = log2(pmax(1,numReads)))
+    mutate(numReads = log2(pmax(1,1+(numReads/scalingFactor))))
     return(outData)
 }
 
 #' ensures that the data is trainable after filtering
 checkFeatures = function(features){
     labels = features$labels
+    trainable = TRUE
     if(sum(labels)==length(labels) | sum(labels)==0){
-    message("Missing presence of both TRUE and FALSE labels.")
-    return(FALSE)
+        message("Missing presence of both TRUE and FALSE labels.")
+        trainable = FALSE
     }
-    if(length(labels)<50){
+    if(length(labels)<1000){
         message("Not enough data points")
-        return(FALSE)
+        trainable = FALSE
+    }
+    if(sum(labels)<50 | sum(!labels)<50){
+        message("Not enough TRUE/FALSE labels")
+        trainable = FALSE
     }
-    return(TRUE)
+    return(trainable)
 }
 
 #' calculates a score based on how likely a read class is full length
-getTranscriptScore = function(rowData){
+getTranscriptScore = function(rowData, defaultModels, fit = TRUE){
     txFeatures = prepareTranscriptModelFeatures(rowData)
     features = dplyr::select(txFeatures,!c(labels))
-    if(checkFeatures(txFeatures)){
-        modelFeatures = features[which(!rowData$novel),]
-        transcriptModel = fit_xgb(modelFeatures, 
-            txFeatures$labels[which(!rowData$novel)])
-        txScore = predict(transcriptModel, as.matrix(features))
-
-        #calculates the FDR for filtering RCs based on wanted precision
+    if(checkFeatures(txFeatures) & fit){
+        ## Multi-Exon
+        indexME = which(!rowData$novel & rowData$numExons>1)
+        transcriptModelME = fitXGBoostModel(
+                    data.train=as.matrix(features[indexME,]),
+                    labels.train=txFeatures$labels[indexME], show.cv=FALSE)
+        txScoreME = predict(transcriptModelME, as.matrix(features))
+
+        ## Single-Exon
+        indexSE = which(!rowData$novel & rowData$numExons==1)
+        transcriptModelSE = fitXGBoostModel(
+                    data.train=as.matrix(features[indexSE,]),
+                    labels.train=txFeatures$labels[indexSE], show.cv=FALSE)
+        txScoreSE = predict(transcriptModelSE, as.matrix(features))
+        txScore = txScoreME
+        txScore[which(rowData$numExons==1)] =
+            txScoreSE[which(rowData$numExons==1)]
         txFDR = calculateFDR(txScore, txFeatures$labels)
```

Comment:
are both SE and ME txScores comparable, or do they need separate adjustment? This might depend on the balance of class labels in each set

Reply:
Do they need separate adjustment, yes and no. Currently I adjust them together which means 0.1 NDR means the ~10% novel isoforms will be a mix of ME and SE ranked by score. If I adjusted them separately an NDR of 0.1 would produce ~10% novel ME isoforms and ~10% novel SE isoforms essentially equating them. The third option is having two separate NDR thresholds for ME and SE which adds complexity. What are your thoughts?